utilities/scripts/python/diff.py

```diff
@@ -0,0 +1,156 @@
+from dataclasses import dataclass
+from typing import Dict
+
+
+@dataclass
+class DiffEntry:
+    """
+    A data class representing an entry in a difference collection.
+    Attributes:
+    val: any
+        The value of the entry.
+    add_or_remove_flag: bool
+        A flag indicating whether the entry should be removed (False) or added (True).
+    """
+
+    val: any
+    add_or_remove_flag: bool
+
+
+@dataclass
+class Diff:
+    """
+    Represents the difference between two dictionaries.
+    """
+    diff: dict
+
+    def __init__(self, expected: dict, provided: dict, strict: bool):
+        if strict:
+            self.diff = __strict_diff__(expected, provided)
+        else:
+            self.diff = __inclusion_diff__(expected, provided)
+
+    def has_diff(self) -> bool:
+        return True if diff else False
+
+    def to_ascii_colored_string(
+        self,
+        obj_name_to_add: str,
+        obj_name_to_remove: str,
+    ) -> str:
+        """
+        Generates an ascii colored string representation of the diff.
+
+        Args:
+            obj_name_to_add (str): The name of the object to add.
+            obj_name_to_remove (str): The name of the object to remove.
+        """
+
+        def __impl__(
+            diff: dict,
+            obj_name_to_add: str,
+            obj_name_to_remove: str,
+            ident: str = "",
+            path: str = "",
+        ) -> str:
+            result = ""
+            if isinstance(diff, DiffEntry):
+                color = "green" if diff.add_or_remove_flag else "red"
+                minus_or_plus = "+" if diff.add_or_remove_flag else "-"
+                obj_name = (
+                    obj_name_to_add if diff.add_or_remove_flag else obj_name_to_remove
+                )
+
+                result += "\n------\n"
+                result += __add_color__(obj_name, "yellow")
+                result += f"{path}\n"
+                result += __add_color__(f"{minus_or_plus}{ident} {diff.val}", color)
+
+            if isinstance(diff, dict):
+                for key in diff:
+                    result += __impl__(
+                        diff[key],
+                        obj_name_to_add,
+                        obj_name_to_remove,
+                        ident + " ",
+                        path + "\n" + f"{ident}  {key}",
+                    )
+
+            if isinstance(diff, list):
+                for val in diff:
+                    result += __impl__(
+                        val, obj_name_to_add, obj_name_to_remove, ident, path
+                    )
+
+            return result
+
+        return __impl__(self.diff, obj_name_to_add, obj_name_to_remove)
+
+
+def __inclusion_diff__(expected: dict, provided: dict) -> dict:
```

Comment:
Is there a reason for using double underscores for this? 

Usually these are reserved for special methods. If we want these to be "private", maybe we could use single underscores? `_inclusion_diff` instead of `__inclusion_diff__`.

Reply:
No, I just put for any reason 😊
Will put with one `_`